File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/LmsFeatureModelsChecklist.kt
package com.lextures.android.core.lms

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

// region Course checklist (CC.9)

@Serializable
enum class ChecklistStatus {
    @SerialName("done") Done,
    @SerialName("todo") Todo,
    @SerialName("in_progress") InProgress,
    @SerialName("unknown") Unknown,
    @SerialName("not_applicable") NotApplicable,
}

@Serializable
enum class ChecklistTier {
    @SerialName("essential") Essential,
    @SerialName("recommended") Recommended,
}

@Serializable
enum class ChecklistDismissReason {
    @SerialName("not_applicable") NotApplicable,
    @SerialName("done_elsewhere") DoneElsewhere,
    @SerialName("disagree") Disagree,
    @SerialName("later") Later,
    @SerialName("other") Other,
    ;

    val wireValue: String
        get() = when (this) {
            NotApplicable -> "not_applicable"
            DoneElsewhere -> "done_elsewhere"
            Disagree -> "disagree"
            Later -> "later"
            Other -> "other"
        }
}

@Serializable
data class ChecklistNavTarget(
    val route: String,
    val anchor: String? = null,
    val entityKey: String? = null,
)

@Serializable
data class ChecklistEvidenceRow(
    val label: String,
    val sublabel: String? = null,
    val status: String = "",
    val target: ChecklistNavTarget? = null,
)

@Serializable
data class ChecklistEvidence(
    val columns: List<String> = emptyList(),
    val rows: List<ChecklistEvidenceRow> = emptyList(),
    val truncatedAt: Int? = null,
)

@Serializable
data class ChecklistDismissal(
    val dismissedAt: String = "",
    val byUserId: String = "",
    val byDisplayName: String = "",
    val reason: String = "",
    val note: String = "",
)

@Serializable
data class ChecklistProgress(
    val done: Int = 0,
    val total: Int = 0,
)

/** Optional assisted-fix primary action (CC.10 FR-5). Unknown kinds render nothing. */
@Serializable
data class ChecklistAction(
    val kind: String = "",
    val labelKey: String = "",
    val label: String = "",
    val endpoint: String = "",
    val requiresAi: Boolean = true,
)

@Serializable
data class ChecklistItem(
    val id: String,
    val titleKey: String = "",
    val title: String = "",
    val whyKey: String = "",
    val why: String = "",
    val tier: ChecklistTier = ChecklistTier.Recommended,
    val status: String = "unknown",
    val detail: String? = null,
    val progress: ChecklistProgress? = null,
    val sources: List<String> = emptyList(),
    val helpRef: String? = null,
    val target: ChecklistNavTarget? = null,
    val evidence: ChecklistEvidence? = null,
    val action: ChecklistAction? = null,
    val dismissal: ChecklistDismissal? = null,
)

@Serializable
data class ChecklistCategory(
    val id: String,
    val titleKey: String = "",
    val title: String = "",
    val items: List<ChecklistItem> = emptyList(),
)

@Serializable
data class CourseChecklistSummary(
    val outstandingEssential: Int = 0,
    val outstandingTotal: Int = 0,
    val done: Int = 0,
    val total: Int = 0,
    val dismissed: Int = 0,
    val computedAt: String = "",
    val stale: Boolean = false,
)

@Serializable
data class CourseChecklist(
    val courseCode: String = "",
    val engineVersion: Int = 0,
    val catalogVersion: String = "",
    val computedAt: String = "",
    val stale: Boolean = false,
    val evidenceTruncated: Boolean = false,
    val summary: CourseChecklistSummary = CourseChecklistSummary(),
    val categories: List<ChecklistCategory> = emptyList(),
    val dismissed: List<ChecklistItem> = emptyList(),
)

@Serializable
data class ChecklistDismissBody(
    val reason: String,
    val note: String = "",
)

// endregion

File: clients/ios/Lextures/Core/LMS/LMSFeatureModelsChecklist.swift
import Foundation

// MARK: - Course checklist (CC.9)

enum ChecklistStatus: String, Codable, Equatable {
    case done
    case todo
    case inProgress = "in_progress"
    case unknown
    case notApplicable = "not_applicable"
}

enum ChecklistTier: String, Codable, Equatable {
    case essential
    case recommended
}

enum ChecklistDismissReason: String, Codable, Equatable, CaseIterable, Identifiable {
    case notApplicable = "not_applicable"
    case doneElsewhere = "done_elsewhere"
    case disagree
    case later
    case other

    var id: String { rawValue }

    var labelKey: String {
        switch self {
        case .notApplicable: return "mobile.checklist.dismissReason.not_applicable"
        case .doneElsewhere: return "mobile.checklist.dismissReason.done_elsewhere"
        case .disagree: return "mobile.checklist.dismissReason.disagree"
        case .later: return "mobile.checklist.dismissReason.later"
        case .other: return "mobile.checklist.dismissReason.other"
        }
    }
}

struct ChecklistNavTarget: Codable, Equatable {
    var route: String
    var anchor: String?
    var entityKey: String?
}

struct ChecklistEvidenceRow: Codable, Equatable, Identifiable {
    var id: String { "\(label)|\(sublabel ?? "")|\(status)" }
    var label: String
    var sublabel: String?
    var status: String
    var target: ChecklistNavTarget?
}

struct ChecklistEvidence: Codable, Equatable {
    var columns: [String]
    var rows: [ChecklistEvidenceRow]
    var truncatedAt: Int?
}

struct ChecklistDismissal: Codable, Equatable {
    var dismissedAt: String
    var byUserId: String
    var byDisplayName: String
    var reason: String
    var note: String
}

struct ChecklistProgress: Codable, Equatable {
    var done: Int
    var total: Int
}

/// Optional assisted-fix primary action (CC.10 FR-5). Unknown kinds render nothing.
struct ChecklistAction: Codable, Equatable {
    var kind: String
    var labelKey: String
    var label: String
    var endpoint: String
    var requiresAi: Bool?
}

struct ChecklistItem: Codable, Equatable, Identifiable {
    var id: String
    var titleKey: String
    var title: String
    var whyKey: String
    var why: String
    var tier: ChecklistTier
    var status: String
    var detail: String?
    var progress: ChecklistProgress?
    var sources: [String]
    var helpRef: String?
    var target: ChecklistNavTarget?
    var evidence: ChecklistEvidence?
    var action: ChecklistAction?
    var dismissal: ChecklistDismissal?
}

struct ChecklistCategory: Codable, Equatable, Identifiable {
    var id: String
    var titleKey: String
    var title: String
    var items: [ChecklistItem]
}

struct CourseChecklistSummary: Codable, Equatable {
    var outstandingEssential: Int
    var outstandingTotal: Int
    var done: Int
    var total: Int
    var dismissed: Int
    var computedAt: String
    var stale: Bool
}

struct CourseChecklist: Codable, Equatable {
    var courseCode: String
    var engineVersion: Int
    var catalogVersion: String
    var computedAt: String
    var stale: Bool
    var evidenceTruncated: Bool
    var summary: CourseChecklistSummary
    var categories: [ChecklistCategory]
    var dismissed: [ChecklistItem]
}

struct ChecklistDismissBody: Encodable {
    var reason: String
    var note: String
}
